src/trixi_include.jl

```diff
@@ -45,13 +49,63 @@ function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
     if !mpi_isparallel(Val{:MPIExt}())
         @info "You just called `trixi_include`. Julia may now compile the code, please be patient."
     end
-    Base.include(ex -> replace_assignments(insert_maxiters(ex); kwargs...), mod, elixir)
+    Base.include(ex -> mapexpr(replace_assignments(insert_maxiters(ex); kwargs...)),
+                 mod, elixir)
+end
+
+function trixi_include(mod::Module, elixir::AbstractString; kwargs...)
+    trixi_include(identity, mod, elixir; kwargs...)
 end
 
 function trixi_include(elixir::AbstractString; kwargs...)
     trixi_include(Main, elixir; kwargs...)
 end
 
+"""
+    trixi_include_changeprecision(T, [mod::Module=Main,] elixir::AbstractString; kwargs...)
+
+`include` the elixir `elixir` and evaluate its content in the global scope of module `mod`.
+You can override specific assignments in `elixir` by supplying keyword arguments,
+similar to [`trixi_include`](@ref).
+
+The only difference to [`trixi_include`](@ref) is that the precision of floating-point
+numbers in the included elixir is changed to `T`.
+More precisely, the package [ChangePrecision.jl](https://github.com/JuliaMath/ChangePrecision.jl)
+is used to convert all `Float64` literals, operations like `/` that produce `Float64` results,
+and functions like `ones` that return `Float64` arrays by default, to the desired type `T`.
+See the documentation of ChangePrecision.jl for more details.
+
+The purpose of this function is to conveniently run a full simulation with `Float32`,
+which is orders of magnitude faster on most GPUs than `Float64`, by just including
+the elixir with `trixi_include_changeprecision(Float32, elixir)`.
+Many constructors in the Trixi.jl framework are written in a way that changing all floating-point
+arguments to `Float32` will change the element type to `Float32` as well.
+In TrixiParticles.jl, including an elixir with this macro should be sufficient
+to run the full simulation with single precision.
+"""
+function trixi_include_changeprecision(T, mod::Module, filename::AbstractString; kwargs...)
+    trixi_include(expr -> ChangePrecision.changeprecision(T, replace_trixi_include(T, expr)),
+                  mod, filename; kwargs...)
+end
+
+function trixi_include_changeprecision(T, filename::AbstractString; kwargs...)
+    trixi_include_changeprecision(T, Main, filename; kwargs...)
+end
+
+function replace_trixi_include(T, expr)
+    expr = TrixiBase.walkexpr(expr) do x
+        if x isa Expr
+            if x.head === :call && x.args[1] === :trixi_include
+                x.args[1] = :trixi_include_changeprecision
+                insert!(x.args, 2, :($T))
+            end
+        end
+        return x
+    end
+
+    return expr
+end
```

Comment:
One more problem here. This doesn't work when, for some reason, we would include an elixir that looks like this:
```julia
using TrixiBase

TrixiBase.trixi_include(...)
```

It only works when we directly write the elixir as
```julia
using TrixiBase

trixi_include(...)
```

@ranocha any ideas how to fix this?

Reply:
Can you special case this with something like Symbol("TrixiBase.trixi_include")?

Alternatively, @vchuravy might have an idea how this can be done?

Reply:
I tried, didn't work. Also, we would need to add a special case for each package that re-exports this name.

Reply:
This will be a `GlobalRef(mod, name)`.

So `x.args[1] isa GlobalRef` and then check the name and replace it.

Reply:
Doesn't work. `x.args[1] isa GlobalRef` is `false`.

Reply:
What is `x.args[1]` when it is the case that doesn't work?

Reply:
`x.head = :call` and `x.args[1] = :(TrixiParticles.trixi_include)`

Reply:
Why do you want to write `TrixiParticles.trixi_include` instead of `trixi_include`?

Reply:
```
julia> dump(:(TrixiParticles.trixi_include))
Expr
  head: Symbol .
  args: Array{Any}((2,))
    1: Symbol TrixiParticles
    2: QuoteNode
      value: Symbol trixi_include
```

Ah so this is before resolution, you will have the issue that someone might also write: 

```
Main.TrixiParticles.trixi_include
```

in any arbitrary depth and length.

Reply:
So is there a simple solution for this? Or do we just ignore this case?

Reply:
The most simple hack would be to check specifically then whether we have `Trixi.trixi_include` or `TrixiParticles.trixi_include`. Alternatively, you could try

```suggestion
function replace_trixi_include(T, expr)
    expr = TrixiBase.walkexpr(expr) do x
        if x isa Expr
            # allow `trixi_include(...)` and `Trixi.trixi_include(...)` etc.
            split_arg = split(string(x.args[1]), '.')
            if x.head === :call && split_arg[end] == "trixi_include"
                arg_pefix = join(split_args[begin:(end - 1)], '.')
                x.args[1] = Symbol(arg_pefix * "trixi_include_changeprecision")
                insert!(x.args, 2, :($T))
            end
        end
        return x
    end

    return expr
end
```

Reply:
```
julia> TrixiBase.trixi_include_changeprecision(Float32, "../TrixiBase.jl/test1.jl")
[ Info: You just called `trixi_include`. Julia may now compile the code, please be patient.
ERROR: LoadError: UndefVarError: `TrixiBase.trixi_include_changeprecision` not defined in `Main`
Suggestion: check for spelling errors or missing imports.
```

Reply:
But it looks like `TrixiBase.trixi_include_changeprecision` is defined... Does your test file create a module? How does it import `trixi_include`?

Reply:
Idk. The test file just contains this single line:
```julia
TrixiBase.trixi_include("test2.jl")
```
And the error message says we're in `Main`.

Reply:
Or should we merge it as is and discuss this in an issue? I don't feel like wasting more time on this edge case.